### PR TITLE
Yul generation of "require" and "assert"

### PR DESCRIPTION
**Pull request by Marenz**
This is a first draft of the yul code generation of "require" and "assert".
Particular the bit with string parameters for require is shaky and untested.
I am also not sure how to actually test that part.

Part of #8343.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Perhaps move this to its own function?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
            m_code << Whiskers("if iszero(<cond>) { <invalidOrRevert> }")
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
        u256 const errHash =
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Don't we have that in the context?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Shouldn't use `0x40` literally, but either call a function or retrieve it from the utils / context.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Unfortunately not, no. Not sure it's worth having either, as it can easily be created on demand

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Ok!

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `6e6f4b83ae34b3d06bda4d05950544fbb31eb0cd`:
```
Coding style error:
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:355: The code below is completely untested as we don't support StringLiterals yet
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:356: TODO */

```
Please check that your changes are working as intended.

**Comment by Marenz** [comment]
Updated:
* Moved code into separate function
* Added test for assert
* added comment about the stringliteral part

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: We somehow have to enforce that this variable won't clash with another one.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I think it is more than fine to use up 3 more bytes for this one.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: We could have a convention to use `l_` for those?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Another option would be to extract this into a yul function right away...

**Comment by Marenz** [comment]
turned it into functions now. Also added a test that has both `require` and `assert` to make sure the generated functions won't clash

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: I think you have to "escape" `require` instead of `assert` - but probably escaping both would be a good idea. What about `x_helper`?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Hm, I'm not 100% sure, but it would probably be better to also encode `_messageType->identifier()` into the function name.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Rather `byteSize`?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Does not really need to be a local variable.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
			{TypeProvider::stringMemory}
```

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: `if iszero(condition)` is missing :)

Comment (Marenz) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Woops

**Comment by leonardoalt** [comment]
I think it'd be good to have tests like
```
...
assert(false);
require(true);
```
and
```
...
require(false);
assert(true);
```
ensuring they fail.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Please group `codegen` includes together.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Is there a reason why this is done in `visit(FunctionCall)` instead of `endVisit(FunctionCall)`? It could avoid calling `accept` explicitly.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:321]: Could assert that `size() <= 2`

File: libsolidity/codegen/YulUtilFunctions.h
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]:
```suggestion
	// `assert` or `require` call.
```

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Shouldn't this be `!assert || !_messageType`?

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
			("invalidOrRevert", _assert ? "invalid()" : "revert(0, 0)")
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: No specific reason other than I wanted to use the existing switch. I can change it

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:321]: True. I imitated the assert I found in the EVM implementation assuming the original author had their reason for doing it that way.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (Marenz) [reply libsolidity/codegen/YulUtilFunctions.cpp]: This is indeed wrong and will be fixed with the next push

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This won't be done in this PR just yet because @chriseth already did this change in another PR

Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Alright, it's the end.. visit now.

**Comment by Marenz** [comment]
Ready for Review

File: test/libsolidity/semanticTests/viaYul/require.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/viaYul/require.sol:15]: I think we should still have the test and then auto-update the expectation once we can check for it.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		solAssert(arguments.size() <= 2, "Expected no more than two parameters for require/assert");
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		string requireOrAssertFunction = m_utils.requireOrAssertFunction(
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		m_code << move(requireOrAssertFunction) << "(" << m_context.variable(*arguments[0]);
```

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Another option would be to not store it in a variable in the first place.

File: test/libsolidity/semanticTests/viaYul/require.sol
Comment (Marenz) [reply test/libsolidity/semanticTests/viaYul/require.sol:15]: Unfortunately that is not possible. There is no expectation that passes.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Isn't this a pretty easy task for the basic optimizer in any modern c++ compiler?

File: test/libsolidity/semanticTests/viaYul/require.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/viaYul/require.sol:15]: You mean just `FAILURE` does not pass?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: It might, but it also might not. You are telling the compiler here to perform a copy, so it will allocate new memory. The optimizer might or might not be able to remove that allocation.

File: test/libsolidity/semanticTests/viaYul/require.sol
Comment (Marenz) [reply test/libsolidity/semanticTests/viaYul/require.sol:15]: Exactly. Because the same check also checks that the returned data has length 0

**Comment by chriseth** [comment]
If we don't want to wait for the missing isoltest feature before we merge this, then please add a solUnimplemented assert for the two-parameter case.

**Comment by Marenz** [comment]
Made the requested adjustments

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:128]: Please also add a solUnimplementedAssert here so we don't accidentally call it from somewhere else.

Comment (Marenz) [reply libsolidity/codegen/YulUtilFunctions.cpp:128]: I move it here, i don't think we need one at both places

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp:128]: Is this code now duplicated? Couldn't it just use `conversionFunctionSpecial`?

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:128]: The code here is encoding the event signature - conversionFunctionSpecial converts a string literal to a memory string.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:322]: I think this will generate invalid code if a string literal is used, but it is deactivated anyway, so it's fine.

**Review by chriseth — APPROVED** [review]

**Pull request merged**
